server/src/main/java/org/apache/druid/server/coordinator/duty/KillStalePendingSegments.java

```diff
@@ -0,0 +1,139 @@
+
+package org.apache.druid.server.coordinator.duty;
+
+import com.google.common.collect.ImmutableList;
+import com.google.inject.Inject;
+import org.apache.druid.common.guava.FutureUtils;
+import org.apache.druid.indexer.TaskStatusPlus;
+import org.apache.druid.java.util.common.DateTimes;
+import org.apache.druid.java.util.common.logger.Logger;
+import org.apache.druid.rpc.indexing.OverlordClient;
+import org.apache.druid.server.coordinator.DruidCoordinatorRuntimeParams;
+import org.apache.druid.server.coordinator.stats.Dimension;
+import org.apache.druid.server.coordinator.stats.RowKey;
+import org.apache.druid.server.coordinator.stats.Stats;
+import org.joda.time.DateTime;
+import org.joda.time.Interval;
+import org.joda.time.Period;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * Duty to kill stale pending segments which are not needed anymore. Pending segments
+ * are created when appending realtime or batch tasks allocate segments to build
+ * incremental indexes. Under normal operation, these pending segments get committed
+ * when the task completes and become regular segments. But in case of task failures,
+ * some pending segments might be left around and cause clutter in the metadata store.
+ * <p>
+ * While cleaning up, this duty ensures that the following pending segments are
+ * retained for at least {@link #DURATION_TO_RETAIN}:
+ * <ul>
+ * <li>Pending segments created by any active task (across all datasources)</li>
+ * <li>Pending segments created by the latest completed task (across all datasources)</li>
+ * </ul>
+ */
+public class KillStalePendingSegments implements CoordinatorDuty
+{
+  private static final Logger log = new Logger(KillStalePendingSegments.class);
+  private static final Period DURATION_TO_RETAIN = new Period("P1D");
+
+  private final OverlordClient overlordClient;
+
+  @Inject
+  public KillStalePendingSegments(OverlordClient overlordClient)
+  {
+    this.overlordClient = overlordClient;
+  }
+
+  @Override
+  public DruidCoordinatorRuntimeParams run(DruidCoordinatorRuntimeParams params)
+  {
+    final Set<String> killDatasources = new HashSet<>(
+        params.getUsedSegmentsTimelinesPerDataSource().keySet()
+    );
+    killDatasources.removeAll(
+        params.getCoordinatorDynamicConfig()
+              .getDataSourcesToNotKillStalePendingSegmentsIn()
+    );
+
+    final DateTime minCreatedTime = getMinCreatedTimeToRetain();
+    for (String dataSource : killDatasources) {
+      int pendingSegmentsKilled = FutureUtils.getUnchecked(
+          overlordClient.killPendingSegments(
+              dataSource,
+              new Interval(DateTimes.MIN, minCreatedTime)
+          ),
+          true
+      );
+      if (pendingSegmentsKilled > 0) {
+        log.info(
+            "Killed [%d] pendingSegments created before [%s] for datasource[%s].",
+            pendingSegmentsKilled, minCreatedTime, dataSource
+        );
+        params.getCoordinatorStats().add(
```

Comment:
Why not add the stats to emit metrics even when the number of pending segments killed is 0?

Reply:
Just avoiding zero value metric emissions for now, as this is a new metric.